pkg/sql/sql_activity_update_job.go

```diff
@@ -0,0 +1,651 @@
+	"github.com/cockroachdb/errors"
+)
+
+// Enabled enables the stats activity flush job.
+var Enabled = settings.RegisterBoolSetting(
+	settings.SystemOnly,
+	"sql.stats.activity.flush.enabled",
+	"enable the flush to the system statement and transaction activity tables", true)
+
+// FlushInterval defines the interval to flush the stats
+// to the activity tables.
+var FlushInterval = settings.RegisterDurationSetting(
+	settings.SystemOnly,
+	"sql.stats.activity.flush.interval",
+	"the frequency at which the statistics are flushed to activity tables",
+	15*time.Minute,
+	settings.NonNegativeDurationWithMaximum(24*time.Hour),
+)
+
+// SQLStatsActivityAggregationInterval is the cluster setting that controls the aggregation
+// interval for stats when we flush to disk.
+var SQLStatsActivityAggregationInterval = settings.RegisterDurationSetting(
+	settings.TenantWritable,
+	"sql.stats.activity.aggregation.interval",
+	"the interval at which we aggregate SQL activity statistics upon flush, "+
+		"this value must be greater than or equal to sql.stats.activity.flush.interval",
+	time.Hour,
+	settings.NonNegativeDurationWithMaximum(time.Hour*24),
+)
+
+// SQLStatsActivityTopLimit is the cluster setting that controls the number of
+// rows SELECTed to be inserted into the activity tables
+var SQLStatsActivityTopLimit = settings.RegisterIntSetting(
+	settings.TenantWritable,
+	"sql.stats.activity.top.limit",
+	"the limit for the top number of statistics to be flushed to the, "+
+		"the activity tables.",
+	500,
+	settings.NonNegativeInt,
+)
+
+// SQLStatsActivityMaxPersistedRows specifies maximum number of rows that will be
+// retained in system.statement_statistics and system.transaction_statistics.
+// Defaults computed 500(top limit)*6(num columns)*24(hrs)*3(days)=216000
+// to give a minimum of 3 days of history. It was rounded down to 200k to
+// give an even number. The top 500 are likely the same for several columns
+// ,so it should still give 3 days of history.
+var SQLStatsActivityMaxPersistedRows = settings.RegisterIntSetting(
+	settings.TenantWritable,
+	"sql.stats.activity.persisted_rows.max",
+	"maximum number of rows of statement and transaction"+
+		" activity that will be persisted in the system tables.",
+	200000, /* defaultValue*/
+	settings.NonNegativeInt,
+).WithPublic()
+
+const numberOfTopColumns = 6
+
+type sqlActivityUpdateJob struct {
+	job *jobs.Job
+}
+
+// Resume implements the jobs.sqlActivityUpdateJob interface.
+// The SQL activity job runs AS a forever-running background job.
+// and runs the SqlActivityUpdater according to sql.stats.activity.flush.interval.
+func (j *sqlActivityUpdateJob) Resume(ctx context.Context, execCtxI interface{}) (jobErr error) {
+	log.Infof(ctx, "starting sql stats activity flush job")
+	execCtx := execCtxI.(JobExecContext)
+	stopper := execCtx.ExecCfg().DistSQLSrv.Stopper
+	settings := execCtx.ExecCfg().Settings
+
+	for {
+		sampleInterval := FlushInterval.Get(&settings.SV)
+		select {
```

Comment:
You probably want to call `MarkIdle` before going to sleep here, otherwise you'll prevent serverless pod scale-down while you're sleeping. Arguable whether you want to set it back to false while you're inside the active lines below -- it probably doesn't have any effect (the metrics collection period is such that in all likelihood nobody would ever notice that you went to idle=false for the maybe second that you'll spend running queries?)

Reply:
In the case of keyviz, sounds like a bug.

In the case of stream producer: that job is only ever run by system tenants so it doesn't really matter if it indicates idleness or not since you don't scale down system tenant nodes on idleness.